lib/src/pubspec_utils.dart

```diff
@@ -0,0 +1,104 @@
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+import 'package:pub_semver/pub_semver.dart';
+
+import 'package_name.dart';
+import 'pubspec.dart';
+import 'source/hosted.dart';
+
+/// Returns a new [Pubspec] without [original]'s dev_dependencies.
+Pubspec stripDevDependencies(Pubspec original) {
+  ArgumentError.checkNotNull(original, 'original pubspec');
+
+  return Pubspec(
+    original.name,
+    version: original.version,
+    sdkConstraints: original.sdkConstraints,
+    dependencies: original.dependencies.values,
+    devDependencies: [], // explicitly give empty list, to prevent lazy parsing
+    dependencyOverrides: original.dependencyOverrides.values,
+  );
+}
+
+/// Returns a new [Pubspec] without [original]'s dependency_overrides.
+Pubspec stripDependencyOverrides(Pubspec original) {
+  ArgumentError.checkNotNull(original, 'original pubspec');
+
+  return Pubspec(
+    original.name,
+    version: original.version,
+    sdkConstraints: original.sdkConstraints,
+    dependencies: original.dependencies.values,
+    devDependencies: original.devDependencies.values,
+    dependencyOverrides: [],
+  );
+}
+
+/// Returns new pubspec with the same dependencies as [original] but with the
+/// upper bounds of the constraints removed.
+///
+/// If [upgradeOnly] is provided, only the packages whose names are in
+/// [upgradeOnly] will have their upper bounds removed.
+Pubspec stripVersionUpperBounds(Pubspec original, {List<String> upgradeOnly}) {
+  ArgumentError.checkNotNull(original, 'original pubspec');
+  upgradeOnly ??= [];
+
+  List<PackageRange> _stripUpperBounds(
+    Map<String, PackageRange> constrained,
+  ) {
+    final result = <PackageRange>[];
+
+    for (final name in constrained.keys) {
+      final packageRange = constrained[name];
+      var unconstrainedRange = packageRange;
+
+      /// We only need to remove the upper bound if it is a hosted package.
+      if (packageRange.source is HostedSource &&
+          (upgradeOnly.isEmpty || upgradeOnly.contains(packageRange.name))) {
+        unconstrainedRange = PackageRange(
+            packageRange.name,
+            packageRange.source,
+            stripUpperBound(packageRange.constraint),
+            packageRange.description,
+            features: packageRange.features);
+      }
+      result.add(unconstrainedRange);
+    }
+
+    return result;
+  }
+
+  return Pubspec(
+    original.name,
+    version: original.version,
+    sdkConstraints: original.sdkConstraints,
+    dependencies: _stripUpperBounds(original.dependencies),
+    devDependencies: _stripUpperBounds(original.devDependencies),
+    dependencyOverrides: original.dependencyOverrides.values,
+  );
+}
+
+/// Removes the upper bound of [constraint]. If [constraint] is the
+/// empty version constraint, an empty version range will be returned.
+@visibleForTesting
+VersionRange stripUpperBound(VersionConstraint constraint) {
```

Comment:
Can't we do `VersionConstraint`? is there any places where it needs to be a `VersionRange`?
```suggestion
VersionConstraint stripUpperBound(VersionConstraint constraint) {
```

For testing we can just do `as VersionRange`.